content/blog/its-time-to-try-out-pixi/contents.lr

```diff
@@ -0,0 +1,1048 @@
+```toml
+# FILE: pyproject.toml
+[project]
+name = "pixi-cuda-environment"
+version = "0.1.0"
+description = "Add a short description here"
+authors = [{name = "Eric Ma", email = "e************@gmail.com"}]
+requires-python = ">= 3.11"
+dependencies = ["pandas>=2.2.2,<2.3"] # <-- NOTE: stuff added via `pixi add --pypi` goes here!
+
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[tool.pixi.project]
+channels = ["conda-forge"]
+platforms = ["osx-arm64", "linux-64"] # <-- NOTE: I modified this manually to include linux-64, and you can add osx-64 if you're on an Intel Mac!
+
+[tool.pixi.pypi-dependencies]
+pixi-cuda-environment = { path = ".", editable = true } # <-- NOTE: this ensures that we never have to run `pip install -e .` ourselves!
+
+[tool.pixi.tasks]
+
+[tool.pixi.dependencies] # <-- NOTE: stuff added via `pixi add` goes here!
+ruff = ">=0.5.5,<0.6"
+pytest = ">=8.3.2,<8.4"
+```
+
+### "Activating" an environment
+
+Because there's no centralized conda environment,
+if you need to run commands within the pixi environment
+(which has the same structure as a conda environment),
+you run `pixi shell` rather than `conda activate <env_name>`.
+
+Running `pixi shell` at my terminal results in something that looks like this:
+
+```bash
+❯ which python
+python not found
+
+❯ pixi shell
+ . "/var/folders/lb/fzbrctwd2klcrwsg63vzqj0r0000gn/T/pixi_env_37v.sh"
+
+❯  . "/var/folders/lb/fzbrctwd2klcrwsg63vzqj0r0000gn/T/pixi_env_37v.sh"
+
+(pixi-cuda-environment)
+❯ which python
+/Users/ericmjl/github/incubator/pixi-cuda-environment/.pixi/envs/default/bin/python
+
+(pixi-cuda-environment)
+❯
+```
+
+Notice how _before_ running `pixi shell`, I couldn't execute Python natively in my terminal.
+After all, I went cold turkey and deleted my `mambaforge` installation!
+(There is a twist: you _can_ use `pixi` to install Python globally!
+I address this below.
+Though for the sake of simplicity for now,
+let's assume that a global Python installation doesn't exist.)
+But _after_ running `pixi shell`, I can --
+because the `/Users/ericmjl/github/incubator/pixi-cuda-environment/.pixi/envs/default/bin/python` path
+is now set on my [`PATH` environment variable](https://ericmjl.github.io/essays-on-data-science/software-skills/environment-variables/).
+
+### Adding CUDA dependencies
+
+At this point, NVIDIA hardware is
+the default hardware used by data scientists for GPU-accelerated computation.
+As such, I need to know how to use `pixi` to install GPU-enabled packages.
+I went to my go-to, JAX and PyTorch,
+to figure out how to install the GPU-enabled versions of these packages.
+
+To make this happen, we must distinguish between both environments:
+CPU-only, like on my Mac,
+and GPU-enabled, like on my home lab that runs Ubuntu.
+First, we define the common and CUDA-specific dependencies on CUDA-enabled systems.
+To start, my `pyproject.toml` file changes a bit on the `dependencies` section
+within the `pyproject.toml` file:
```

Comment:
Did you mean to mention `pyproject.toml` twice in this sentence?

Reply:
Great catch, thank you!